### DISCUSSION TITLE
LSP setup for nvim-tree development

### DISCUSSION THREAD

**Discussion by remvn** [nvim-tree/nvim-tree.lua#2901] [Q&A] [answered]
When I was checking for vim api type definition (the one from library) using `vim.lsp.buf.definition`

I ran into this error message (even though it still takes me to the definition of that type)
![image](https://github.com/user-attachments/assets/60a1d87e-94c8-4e1c-b0c8-2c203b0662ac)

and the root directory in :LspInfo changes to the home directory:
![image](https://github.com/user-attachments/assets/246242a0-3d97-4374-98a1-3428a2b133cb)

My `lua_ls` setup is auto configured by `mason_lspconfig` , with all the options left default.
Is there anything I'm missing here?

**Comment by alex-courtis** [comment]
> I ran into this error message (even though it still takes me to the definition of that type)

That's normal. lua-language-server will not start with your home directory as root. It's not clear from the documentation link, however I think that's reasonable as it would scan _every lua file_ under your home.

Why does this happen? It could be scanning up the directory tree for `.luarc.json` files and finds only `~/.luarc.json`.

I'm not familiar with `meson_lspconfig`, however `neovim/nvim-lspconfig` and a bare `vim.lsp.start` call both work. I'll update the wiki.

**Comment by alex-courtis** [comment]
Native, neovim/nvim-lspconfig and hrsh7th/nvim-cmp configurations and samples added to wiki: 
[Lua Language Server](https://github.com/nvim-tree/nvim-tree.lua/wiki/Development#lua-language-server)

I'd be most grateful if you tested those, as well as adding a `mason_lspconfig` section.

**Reply by remvn** [reply, marked as answer]
I believe `mason-lspconfig` is just a wrapper for calling `nvim-lspconfig` default setup automatically [mason-lspconfig docs ](https://github.com/williamboman/mason-lspconfig.nvim/blob/main/doc/server-mapping.md)

I tried disable `mason-lspconfig`  and use your setup from the wiki but still the same behavior 

It's working perfectly **within this project scope** (jumping around, has intellisense for vim's api)
However this error only show up when I try to jump to some type definitions in `luv library`
for example:
![image](https://github.com/user-attachments/assets/963cee64-34c5-4c26-a57a-b8f70a64c51f)

Digging around these library directory I found that they don't have `.luarc.json` but they do have a `.git` file
![image](https://github.com/user-attachments/assets/97c92b70-7812-42e0-9cb1-728380bd9252)

And for some reason... `lua_ls` default setup doesn't include `.git` in `root_dir` even though the docs has it [nvim-lspconfig source](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/lua_ls.lua)

Adding `.git` manually and the error will go away:
```lua
lspconfig.lua_ls.setup({
    root_dir = lspconfig.util.root_pattern(
        ".luarc.json",
        ".luarc.jsonc",
        ".luacheckrc",
        ".stylua.toml",
        "stylua.toml",
        "selene.toml",
        "selene.yml",
        ".git"
    ),
})
```
root directory in `:LspInfo` seems correct
![image](https://github.com/user-attachments/assets/0bd62202-322f-4319-8953-b1177423b6c0)

**Reply by alex-courtis** [reply]
The arch package places them in /usr/lib however it's unclear whether the location causes luals to treat them differently.
```
:  ; yay -Ql lua-language-server | grep luv
lua-language-server /usr/lib/lua-language-server/meta/3rd/luv/
lua-language-server /usr/lib/lua-language-server/meta/3rd/luv/config.json
lua-language-server /usr/lib/lua-language-server/meta/3rd/luv/library/
lua-language-server /usr/lib/lua-language-server/meta/3rd/luv/library/errors.lua
...
```

You could try putting a `.luarc.json` in there.

I've got a "standard one" `~/.config/nvim/.luarc.json` which is also linked at `~/.local/share/nvim/site/pack/packer/start/.luarc.json` so that I can hack on other plugins on the runtime path.

``~/.local/share/nvim/.luarc.json` seems like a good balance, I might try that out.

**Reply by alex-courtis** [reply]
> And for some reason... `lua_ls` default setup doesn't include `.git` in `root_dir` even though the docs has it [nvim-lspconfig source](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/lua_ls.lua)

This is very strange. Other configurations do explicitly specify `.git` but lua_ls does not.

It looks like there is extra root logic for some configurations... like finding `.git`
```lua
      root = util.root_pattern 'lua/'(fname)
      if root then
        return root
      end
      return util.find_git_ancestor(fname)
```

Experimenting with a new repo did stop at `.git` with the default config, but not for other repos as you found. It seems that explicitly adding `.git` is the best course of action.

None of this is documented or clear, very disappointing. Thanks for digging, now I know and understand limitations.

**Reply by remvn** [reply]
> Experimenting with a new repo did stop at .git with the default config, but not for other repos as you found. It seems that explicitly adding .git is the best course of action.

Sounds great! Glad that neovim now comes with vim's api types.
